File: server/src/main/java/com/hospitaltender/server/service/OfficialBidService.java
package com.hospitaltender.server.service;

import com.hospitaltender.server.dto.request.CreateOfficialBidRequest;
import com.hospitaltender.server.dto.response.OfficialBidResponse;

import java.util.List;

public interface OfficialBidService {

    OfficialBidResponse create(CreateOfficialBidRequest request);

    OfficialBidResponse getById(Long id);

    List<OfficialBidResponse> getByTenderItemId(Long tenderItemId);

    List<OfficialBidResponse> getValidBidsByTenderItemId(Long tenderItemId);

    void delete(Long id);

    /**
     * Evaluates all bids for a tender item:
     * - Rejects bids that exceed the approximate cost
     * - Selects the lowest valid bid as the winner
     */
    OfficialBidResponse evaluateBids(Long tenderItemId);

    /**
     * Manually reject a bid with a reason
     */
    OfficialBidResponse rejectBid(Long bidId, String reason);
}

File: ui/hospital-system-ui/src/services/official-bid.service.ts
import api from './api'
import { OfficialBidResponse, CreateOfficialBidRequest } from '@/types'

export const officialBidService = {
  getByTenderItemId: (tenderItemId: number) =>
    api.get<OfficialBidResponse[]>(`/official-bids/tender-item/${tenderItemId}`).then((res) => res.data),

  getValidByTenderItemId: (tenderItemId: number) =>
    api.get<OfficialBidResponse[]>(`/official-bids/tender-item/${tenderItemId}/valid`).then((res) => res.data),

  getById: (id: number) =>
    api.get<OfficialBidResponse>(`/official-bids/${id}`).then((res) => res.data),

  create: (data: CreateOfficialBidRequest) =>
    api.post<OfficialBidResponse>('/official-bids', data).then((res) => res.data),

  delete: (id: number) =>
    api.delete(`/official-bids/${id}`),

  evaluate: (tenderItemId: number) =>
    api.post<OfficialBidResponse>(`/official-bids/tender-item/${tenderItemId}/evaluate`).then((res) => res.data),

  reject: (id: number, reason: string) =>
    api.patch<OfficialBidResponse>(`/official-bids/${id}/reject`, null, { params: { reason } }).then((res) => res.data),
}
